[pt1][quant] Add the default_weight_observer for the dynamic quantization path

jianyuh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24231 [pt1][quant] Add the default_weight_observer for the dynamic quantization path**

As suggested in https://github.com/pytorch/pytorch/pull/23128#discussion_r309528932, we will add a default weight observer for the dynamic quantization path.

We need to move `observer` and `qconfig` to a separate namespace.

Differential Revision: [D16781092](https://our.internmc.facebook.com/intern/diff/D16781092/)

jianyuh committed:
[WIP][pt1][quant] Add the default_weight_observer for the dynamic quantization path

As suggested in https://github.com/pytorch/pytorch/pull/23128#discussion_r309528932, we will add a default weight observer for the dynamic quantization path.

We need to move `observer` and `qconfig` to a separate namespace.

Differential Revision: [D16781092](https://our.internmc.facebook.com/intern/diff/D16781092/)

jianyuh committed:
Update on "[WIP][pt1][quant] Add the default_weight_observer for the dynamic quantization path"


As suggested in https://github.com/pytorch/pytorch/pull/23128#discussion_r309528932, we will add a default weight observer for the dynamic quantization path.

We need to move `observer` and `qconfig` to a separate namespace.

Differential Revision: [D16781092](https://our.internmc.facebook.com/intern/diff/D16781092/)

jianyuh committed:
Update on "[WIP][pt1][quant] Add the default_weight_observer for the dynamic quantization path"


As suggested in https://github.com/pytorch/pytorch/pull/23128#discussion_r309528932, we will add a default weight observer for the dynamic quantization path.

We need to move `observer` and `qconfig` to a separate namespace.

Differential Revision: [D16781092](https://our.internmc.facebook.com/intern/diff/D16781092/)

jianyuh committed:
Update on "[WIP][pt1][quant] Add the default_weight_observer for the dynamic quantization path"


As suggested in https://github.com/pytorch/pytorch/pull/23128#discussion_r309528932, we will add a default weight observer for the dynamic quantization path.

We need to move `observer` and `qconfig` to a separate namespace.

Differential Revision: [D16781092](https://our.internmc.facebook.com/intern/diff/D16781092/)

raghuramank100 commented on torch/nn/quantized/dynamic/modules/linear.py:
```diff
@@ -62,7 +62,12 @@ def from_float(cls, mod):
+        if mod.qconfig is not None and mod.qconfig.weight() is not None:
+            weight_observer = mod.qconfig.weight()
+        else:
+            from torch.quantization.observer import default_weight_observer
```
This works, but for clarity, please move the import to the beginning of the file and do:
```
   
def from_float(cls, mod):
...
        if mod.qconfig is not None and mod.qconfig.weight() is not None:
            weight_observer = mod.qconfig.weight()
        else:
            weight_observer = default_weight_observer
...
```

jianyuh replied:
The problem is the circular import issues in Python: https://github.com/pytorch/pytorch/pull/23128#discussion_r310745665. The current solution is to postpone the import of `torch.quantization.observer` until you need it: https://stackoverflow.com/a/10027101/3234205

jianyuh replied:
The ideal solution is to move the observer and qconfig into a separate submodule (https://github.com/pytorch/pytorch/pull/23128#discussion_r311383406). However, it requires a new namespace, which involves many changes. I had discussions with @jerryzh168 and @zafartahirov for this issue.

zafartahirov replied:
This is a valid workaround for the circular dependencies. I think we should add a comment though -- in case some future eng decides to move it out

jianyuh replied:
Added a comment here.

raghuramank100 commented on torch/nn/quantized/dynamic/modules/linear.py:
```diff
@@ -62,7 +62,12 @@ def from_float(cls, mod):
         assert type(mod) == NNLinear, 'nn.quantized.dynamic.Linear.from_float only works for nn.Linear'
         assert hasattr(mod, 'qconfig'), 'Input float module must have qconfig defined'
-        weight_observer = mod.qconfig.weight()
+        if mod.qconfig is not None and mod.qconfig.weight() is not None:
```
Please add a test exercising from_float on a module with and without qconfig. (i.e a plain nn.module)

jianyuh replied:
Will add the unit test for this.

jianyuh replied:
Added the unit test.

raghuramank100 reviewed (changes requested):
Please see comments, thanks

jianyuh committed:
Update on "[pt1][quant] Add the default_weight_observer for the dynamic quantization path"


As suggested in https://github.com/pytorch/pytorch/pull/23128#discussion_r309528932, we will add a default weight observer for the dynamic quantization path.

We need to move `observer` and `qconfig` to a separate namespace.

Differential Revision: [D16781092](https://our.internmc.facebook.com/intern/diff/D16781092/)

zafartahirov commented:
LGTM, but be mindful of the comments about testing the `from_float` by @raghuramank100

jianyuh commented:
> LGTM, but be mindful of the comments about testing the `from_float` by @raghuramank100

Thanks! Will add the unit test for that.

jianyuh committed:
Update on "[pt1][quant] Add the default_weight_observer for the dynamic quantization path"


As suggested in https://github.com/pytorch/pytorch/pull/23128#discussion_r309528932, we will add a default weight observer for the dynamic quantization path.

We need to move `observer` and `qconfig` to a separate namespace.

Differential Revision: [D16781092](https://our.internmc.facebook.com/intern/diff/D16781092/)

jianyuh committed:
Update on "[pt1][quant] Add the default_weight_observer for the dynamic quantization path"


As suggested in https://github.com/pytorch/pytorch/pull/23128#discussion_r309528932, we will add a default weight observer for the dynamic quantization path.

We need to move `observer` and `qconfig` to a separate namespace.

Differential Revision: [D16781092](https://our.internmc.facebook.com/intern/diff/D16781092/)
